### PR TITLE
Moved Egamma TensorFlow sessions to global cache

### PR DESCRIPTION
**Pull request by valsdav**
#### PR description:

Solves #46494.  Related to #46040 

`EgammaDNNHelper` was storing TF graphs globally by job, but TF sessions were owned my the `GSFElectronProducer` and `GEDPhotonProducers` by stream.
This PR moves the TFSessions in the `EgammaDNNHelper`, making them global by job.

#### PR validation:
Purely technical PR, no changes expected.

### CODE REVIEW

**Comment by valsdav** [comment]
assign ml

**Comment by valsdav** [comment]
please test

**Comment by valsdav** [comment]
please test

**Comment by valsdav** [comment]
please abort test

**Comment by valsdav** [comment]
please test

**Comment by valsdav** [comment]
please test

**Comment by valsdav** [comment]
Hi @makortel, I would like to reproduce the memory profile of #46040 to compare the memory usage. Would you have any script to share to reproduce the same igProf memory profile that's in the issue? Thanks!

**Comment by makortel** [comment]
Below is an untested recipe to reproduce what I did for https://github.com/cms-sw/cmssw/issues/46494
```sh
# Set up developer area
cmssw-el7
export SCRAM_ARCH=slc7_amd64_gcc12
cmsrel CMSSW_14_0_16
cd CMSSW_14_0_16/src
cmsenv

# Set up the configuration
cp /eos/home-c/cmst0/public/PausedJobs/Run2024G/maxPSS/PromptReco_Run386319_Muon1/job/WMTaskSpace/cmsRun1/PSet.p* .
xrdcp root://eoscms.cern.ch//eos/cms/tier0/store/data/Run2024H/Muon1/RAW/v1/000/386/319/00000/27604955-1005-455b-90eb-8195e2f02cb8.root .
cat >>PSet.py <<EOF
process.source.fileNames = ["file:27604955-1005-455b-90eb-8195e2f02cb8.root"]
process.maxEvents.input = 20
process.options.numberOfThreads = 4
process.options.numberOfStreams = 4
process.IgProfService = cms.Service('IgProfService',
    reportFirstEvent = cms.untracked.int32(0),
    reportEventInterval = cms.untracked.int32(10),
    reportToFileAtPostEvent     = cms.untracked.string('| gzip -c > 4th_4st_07.mem.%I.gz')
)
EOF

# Run IgProf memory profiler
igprof -d -t cmsRunGlibC -z -mp -o test_07.mem.gz cmsRunGlibC PSet.py > out.txt 2>&1
igprof-analyse --sqlite -d -v -g -r MEM_LIVE test_4th_4st_07.mem.20.gz | sqlite3 test_4th_4st_07.20_live.sql3

# Make the memory profile web browseable (needs a web area with cgi-bin access)
cp (which igprof-navigator) <web dir>/cgi-bin
mkdir <web dir>/cgi-bin/data
cp test_4th_4st_07.20_live.sql3 <web dir>/cgi-bin/data
```
Open web browser to `<web dir>/cgi-bin/igprof-navigator/test_4th_4st_07.20_live`.

**Comment by valsdav** [comment]
Thanks a lot! Unfortunately the data file is no more available, can I just use another one?

File: RecoEgamma/EgammaTools/interface/EgammaDNNHelper.h [hunk 1]
```diff
@@ -65,6 +66,7 @@ namespace egammaTools {
     std::vector<uint> nInputs_;
 
     std::vector<std::unique_ptr<const tensorflow::GraphDef>> graphDefs_;
+    std::vector<std::unique_ptr<tensorflow::Session>> sessions_;
```
Comment (Dr15Jones) [inline RecoEgamma/EgammaTools/interface/EgammaDNNHelper.h:69]: Assuming `tensor flow::Session` is _const thread-safe_ this should be `std::vector<std::unique_ptr<const tensorflow::Session>>` as it is used across threads concurrently.

Comment (makortel) [reply RecoEgamma/EgammaTools/interface/EgammaDNNHelper.h:69]: IIRC `tensorflow::Session` interface is not `const`, even if it is thread safe and does not mutate the `Session` object (i.e. in principle the interface could be `const`) :(

Comment (valsdav) [reply RecoEgamma/EgammaTools/interface/EgammaDNNHelper.h:69]: I'm about to push a commit moving this to the `tensorflow::SessionCache` (https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/TensorFlow/interface/TensorFlow.h#L189) which is returning `const Session*` .

**Comment by valsdav** [comment]
> Below is an untested recipe to reproduce what I did for #46494
> 
> ```shell
> # Set up developer area
> cmssw-el7
> export SCRAM_ARCH=slc7_amd64_gcc12
> cmsrel CMSSW_14_0_16
> cd CMSSW_14_0_16/src
> cmsenv
> 
> # Set up the configuration
> cp /eos/home-c/cmst0/public/PausedJobs/Run2024G/maxPSS/PromptReco_Run386319_Muon1/job/WMTaskSpace/cmsRun1/PSet.p* .
> xrdcp root://eoscms.cern.ch//eos/cms/tier0/store/data/Run2024H/Muon1/RAW/v1/000/386/319/00000/27604955-1005-455b-90eb-8195e2f02cb8.root .
> cat >>PSet.py <<EOF
> process.source.fileNames = ["file:27604955-1005-455b-90eb-8195e2f02cb8.root"]
> process.maxEvents.input = 20
> process.options.numberOfThreads = 4
> process.options.numberOfStreams = 4
> process.IgProfService = cms.Service('IgProfService',
>     reportFirstEvent = cms.untracked.int32(0),
>     reportEventInterval = cms.untracked.int32(10),
>     reportToFileAtPostEvent     = cms.untracked.string('| gzip -c > 4th_4st_07.mem.%I.gz')
> )
> EOF
> 
> # Run IgProf memory profiler
> igprof -d -t cmsRunGlibC -z -mp -o test_07.mem.gz cmsRunGlibC PSet.py > out.txt 2>&1
> igprof-analyse --sqlite -d -v -g -r MEM_LIVE test_4th_4st_07.mem.20.gz | sqlite3 test_4th_4st_07.20_live.sql3
> 
> # Make the memory profile web browseable (needs a web area with cgi-bin access)
> cp (which igprof-navigator) <web dir>/cgi-bin
> mkdir <web dir>/cgi-bin/data
> cp test_4th_4st_07.20_live.sql3 <web dir>/cgi-bin/data
> ```
> 
> Open web browser to `<web dir>/cgi-bin/igprof-navigator/test_4th_4st_07.20_live`.

Hi @makortel, I managed to replicate the recipe.  There was a crash at the end of the processing but it seems that the igprof report has been filled. Unfortunately my cgi-bin setup has some problem (it was working in the past). Would you mind copying the file in your folder please? https://dvalsecc.web.cern.ch/cgi-bin/data/test_4th_4st_07.20_live.sql3

**Comment by jfernan2** [comment]
solves https://github.com/cms-sw/cmssw/pull/46494

**Comment by valsdav** [comment]
+ml

Technical PR, moved TF sessions to global cache

**Comment by valsdav** [comment]
@makortel we probably want to backport this to the release used in T0 processing at the moment, isn't it?

**Comment by makortel** [comment]
> Would you mind copying the file in your folder please? https://dvalsecc.web.cern.ch/cgi-bin/data/test_4th_4st_07.20_live.sql3

I can, but I'd need a path to the file on lxplus that I could access (access through `https` results in the `cgi-bin` script being called that fails with internal server error, and I don't have permissions to access the file through EOS)

**Comment by makortel** [comment]
> we probably want to backport this to the release used in T0 processing at the moment, right?

Based on https://github.com/cms-sw/cmssw/issues/46494 the expected memory savings would be about 7 MB for an 8-stream job. It isn't much compared to a full reconstruction job, but it is something nevertheless. My take is that backports would not be essential, but if done, I'd do backport also to 14_0_X to potentially benefit the MC production.

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**